src/pcap/client.go

```diff
@@ -0,0 +1,335 @@
+
+	// perform capture request
+	err := c.ProcessCapture(ctx, endpointRequest, options, cancel)
+	if err != nil {
+		return fmt.Errorf("encountered error during request handling: %w", err)
+	}
+
+	// handle results of capture request
+	cause := context.Cause(ctx)
+	if cause != nil && !errors.Is(cause, context.Canceled) {
+		return fmt.Errorf("finished with error: %w", cause)
+	}
+	return nil
+}
+
+// ProcessCapture takes care of the complete lifecycle for a capture request.
+// It writes the pcap-header to the outputFile, sends the CaptureRequest to the pcap-api
+// and handles the cleanup after the capture is done.
+//
+// It then delegates writing individual packets and logging messages from the api to ReadCaptureResponse.
+// logProgress is called in another goroutine to asynchronously announce on stderr how many bytes were
+// already written to the outputFile.
+func (c *Client) ProcessCapture(ctx context.Context, endpointRequest *EndpointRequest, options *CaptureOptions, cancel context.CancelCauseFunc) error {
+	logger := c.log.With(zap.String(LogKeyHandler, "ProcessCapture"))
+	if endpointRequest == nil {
+		return fmt.Errorf("endpoint request must not be nil: %w", errInvalidPayload)
+	}
+
+	if options == nil {
+		return fmt.Errorf("capture options request must not be nil: %w", errInvalidPayload)
+	}
+	// setup output/pcap-file
+	packetWriter := pcapgo.NewWriter(c.packetFile)
+	err := packetWriter.WriteFileHeader(options.SnapLen, layers.LinkTypeEthernet)
+	if err != nil {
+		return err
+	}
+
+	captureRequest := &CaptureRequest{
+		Operation: &CaptureRequest_Start{
+			Start: &StartCapture{
+				Request: endpointRequest,
+				Options: options,
+			},
+		},
+	}
+
+	c.stream, err = c.Capture(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = c.stream.Send(captureRequest)
+	if err != nil {
+		return err
+	}
+
+	copyWg := &sync.WaitGroup{}
+	copyWg.Add(1)
+	go c.ReadCaptureResponse(c.stream, packetWriter, copyWg, cancel)
+
+	go c.logProgress(ctx)
+
+	// wait for progress to finish
+	<-ctx.Done()
+
+	logger.Info("waiting for copy operation to stop")
+	copyWg.Wait()
+
+	logger.Debug("syncing file to disk")
+	err = c.packetFile.Sync()
+	if err != nil {
+		return err
+	}
+
+	logger.Debug("closing file")
+	err = c.packetFile.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Client) StopRequest() {
+	if c.stream == nil {
+		c.log.Error("client not connected, could not stop")
+		return
+	}
+	if c.stopped {
+		return
+	}
+
+	err := c.stream.SendMsg(MakeStopRequest())
+	if err != nil {
+		c.log.Error("could not stop")
+	}
+	c.stopped = true
+}
+
+// ReadCaptureResponse reads CaptureResponse's from the api in a loop and delegates writing/logging messages & packets to WriteMessage / writePacket.
+//
+// It terminates if an error or clean stop-message is received.
+func (c *Client) ReadCaptureResponse(stream API_CaptureClient, packetWriter *pcapgo.Writer, copyWg *sync.WaitGroup, cancel context.CancelCauseFunc) {
```

Comment:
@maxmoehl could I not also pass in the context and close (cancel w/o error) that context?

Reply:
It's the other way around. The caller wants to be sure that this function stopped after the context was already cancelled. This cannot be achieved with a context because the context does not guarantee that the work is actually cancelled which is why we need a separate control to ensure the work has stopped.

Reply:
The go idiomatic way would (IMO) be a channel since it's only one routine in the wait group. Technically both approaches work just fine.

Reply:
channel it is then! thank you!